quinn-proto/src/range_set/array_range_set.rs

```diff
@@ -0,0 +1,305 @@
+        for range in self.0.iter() {
+            if range.start > x {
+                // We only get here if there was no prior range that contained x
+                return false;
+            } else if range.contains(&x) {
+                return true;
+            }
+        }
+        false
+    }
+
+    pub fn subtract(&mut self, other: &ArrayRangeSet) {
+        // TODO: This can potentially be made more efficient, since the we know
+        // individual ranges are not overlapping, and the next range must start
+        // after the last one finished
+        for range in &other.0 {
+            self.remove(range.clone());
+        }
+    }
+
+    pub fn insert_one(&mut self, x: u64) -> bool {
+        self.insert(x..x + 1)
+    }
+
+    pub fn insert(&mut self, x: Range<u64>) -> bool {
+        let mut result = false;
+
+        if x.is_empty() {
+            // Don't try to deal with ranges where x.end <= x.start
+            return false;
+        }
+
+        let mut idx = 0;
+        while idx != self.0.len() {
+            let range = &mut self.0[idx];
+
+            if range.start > x.end {
+                // The range is fully before this range and therefore not extensible.
+                // Add a new range to the left
+                self.0.insert(idx, x);
+                return true;
+            } else if range.start > x.start {
+                // The new range starts before this range but overlaps.
+                // Extend the current range to the left
+                // Note that we don't have to merge a potential left range, since
+                // this case would have been captured by merging the right range
+                // in the previous loop iteration
+                result = true;
+                range.start = x.start;
+            }
+
+            // At this point we have handled all parts of the new range which
+            // are in front of the current range. Now we handle everything from
+            // the start of the current range
+
+            if x.end <= range.end {
+                // Fully contained
+                return result;
+            } else if x.start <= range.end {
+                // Extend the current range to the end of the new range.
+                // Since it's not contained it must be bigger
+                range.end = x.end;
+
+                // Merge all follow-up ranges which overlap
+                while idx != self.0.len() - 1 {
+                    let curr = self.0[idx].clone();
+                    let next = self.0[idx + 1].clone();
+                    if curr.end >= next.start {
+                        self.0[idx].end = next.end.max(curr.end);
+                        self.0.remove(idx + 1);
+                    } else {
+                        break;
+                    }
+                }
+
+                return true;
+            }
+
+            idx += 1;
+        }
+
+        if x.is_empty() {
+            return result;
+        }
```

Comment:
If `x` is no longer mutable, this branch is redundant to the one at the start of the function.